src/javascript/lib/domstubs.js

```diff
@@ -0,0 +1,268 @@
+          if (this._loopIndex < this._attributeKeys.length) {
+            var name = this._attributeKeys[this._loopIndex++];
+            return ' ' + name + '="' + xmlEncode(node.attributes[name]) + '"';
+          }
+          ++this._state;
+          return '>';
+        case 4:  // Serialize textContent for tspan/style elements.
+          if (node.nodeName === 'svg:tspan' || node.nodeName === 'svg:style') {
+            this._state = 6;
+            return xmlEncode(node.textContent);
+          }
+          ++this._state;
+          this._loopIndex = 0;
+          /* falls through */
+        case 5:  // Serialize child nodes (only for non-tspan/style elements).
+          var value;
+          while (true) {
+            value = this._childSerializer && this._childSerializer.getNext();
+            if (value !== null) {
+              return value;
+            }
+            var nextChild = node.childNodes[this._loopIndex++];
+            if (nextChild) {
+              this._childSerializer = new DOMElementSerializer(nextChild);
+            } else {
+              this._childSerializer = null;
+              ++this._state;
+              break;
+            }
+          }
+          /* falls through */
+        case 6:  // Ending tag.
+          ++this._state;
+          return '</' + node.nodeName + '>';
+        case 7:  // Done.
+          return null;
+        default:
+          throw new Error('Unexpected serialization state: ' + this._state);
+      }
+    },
+  };
+  
+  const document = {
+    childNodes: [],
+  
+    get currentScript() {
+      return { src: '', };
+    },
+  
+    get documentElement() {
+      return this;
+    },
+  
+    createElementNS: function(NS, element) {
+      var elObject = new DOMElement(element);
+      return elObject;
+    },
+  
+    createElement: function(element) {
+      return this.createElementNS('', element);
+    },
+  
+    getElementsByTagName: function(element) {
+      if (element === 'head') {
+        return [this.head || (this.head = new DOMElement('head'))];
+      }
+      return [];
+    },
+  };
+  
+  function Image() {
+    this._src = null;
+    this.onload = null;
+  }
+  Image.prototype = {
+    get src() {
+      return this._src;
+    },
+    set src(value) {
+      this._src = value;
+      if (this.onload) {
+        this.onload();
+      }
+    },
+  };
+  
+  exports.document = document;
+  exports.Image = Image;
+  
+  var exported_symbols = Object.keys(exports);
+  
+  exports.setStubs = function(namespace) {
+    exported_symbols.forEach(function(key) {
+      console.assert(!(key in namespace), 'property should not be set: ' + key);
+      namespace[key] = exports[key];
+    });
+  };
+  exports.unsetStubs = function(namespace) {
+    exported_symbols.forEach(function(key) {
+      console.assert(key in namespace, 'property should be set: ' + key);
+      delete namespace[key];
+    });
+  };
```

Comment:
- do we really need the file in its entirety? surely we just need `document`? 
- attribution would be useful - add some jsdoc to indicate that this was taken from the node example of pdf.js

Reply:
Fair, I have removed the Image.prototype. The rest of the functions/objects: `xmlEncode`, `DOMElement`, `DOMElement.prototype` and `DOMElementSerializer` are part of the attributes of `document`, so I've left it there.